persistence-jpa/javadsl/src/main/java/com/lightbend/lagom/internal/javadsl/persistence/jpa/JpaReadSideImpl.java

```diff
@@ -168,15 +171,61 @@ private void updateOffset(EntityManager entityManager, Offset offset) {
+                        offset, readSideId, statement);
+                Query updateOffsetQuery = entityManager.createNativeQuery(statement);
                 // NOTE: The order of parameters here depends on the order chosen
-                // by Slick, based on the table definition in JdbcOffsetStore.
-                entityManager.createNativeQuery(statement)
-                        .setParameter(1, readSideId)
-                        .setParameter(2, tag.tag())
-                        .setParameter(3, sequenceOffset)
-                        .setParameter(4, timeUuidOffset)
-                        .executeUpdate();
+                // by Slick, based on the table definition in JdbcOffsetStore
+                // and the specific database profile in use.
+                JdbcProfile profile = getSlickDatabaseProfile();
+                if (profile instanceof PostgresDriver) {
+                    postgresqlBindUpdateOffsetQuery(updateOffsetQuery, sequenceOffset, timeUuidOffset);
+                } else {
+                    defaultBindUpdateOffsetQuery(updateOffsetQuery, sequenceOffset, timeUuidOffset);
+                }
+                updateOffsetQuery.executeUpdate();
             }
         }
+
+        private JdbcProfile getSlickDatabaseProfile() {
+            return offsetStore.slick().profile();
+        }
+
+        private Query postgresqlBindUpdateOffsetQuery(Query query, Long sequenceOffset, String timeUuidOffset) {
+            // Slick emulates insert or update on PostgreSQL using this compound statement:
+            // begin;
+            //   update "read_side_offsets"
+            //     set "sequence_offset"=?,"time_uuid_offset"=?
+            //     where "read_side_id"=?
+            //     and "tag"=?;
+            //   insert into "read_side_offsets" ("read_side_id","tag","sequence_offset","time_uuid_offset")
+            //     select ?,?,?,?
+            //     where not exists (
+            //       select 1 from "read_side_offsets" where "read_side_id"=? and "tag"=?
+            //     );
+            // end
+            return query
+                    .setParameter(1, sequenceOffset)
+                    .setParameter(2, timeUuidOffset)
+                    .setParameter(3, readSideId)
+                    .setParameter(4, tag.tag())
+                    .setParameter(5, readSideId)
+                    .setParameter(6, tag.tag())
+                    .setParameter(7, sequenceOffset)
+                    .setParameter(8, timeUuidOffset)
+                    .setParameter(9, readSideId)
+                    .setParameter(10, tag.tag());
+        }
+
+        private Query defaultBindUpdateOffsetQuery(Query query, Long sequenceOffset, String timeUuidOffset) {
+            // H2:
+            // merge into "read_side_offsets" ("read_side_id","tag","sequence_offset","time_uuid_offset")
+            //   values (?,?,?,?)
```

Comment:
I'll update the comment with statements for other databases as I get them working.